python/amici/sbml_import.py

```diff
@@ -2020,6 +1922,131 @@ def _get_str_symbol_identifiers(str_symbol: str) -> tuple:
+        ODEModel object with basic definitions
+
+    :param conservation_laws:
+        List of already known conservation laws
+
+    :returns species_solver:
+        List of species indices which remain later in the ODE solver
+    """
+
+    # decide which species to keep in stoichiometry
+    species_solver = list(range(ode_model.num_states_rdata()))
+
+    # iterate over species, find constant ones
+    for ix in reversed(range(ode_model.num_states_rdata())):
+        if ode_model.state_is_constant(ix):
+            # dont use sym('x') here since conservation laws need to be
+            # added before symbols are generated
+            target_state = ode_model._states[ix].get_id()
+            total_abundance = sp.Symbol(f'tcl_{target_state}', real=True)
+            conservation_laws.append({
+                'state': target_state,
+                'total_abundance': total_abundance,
+                'state_expr': total_abundance,
+                'abundance_expr': target_state,
+            })
+            # mark species to delete from stoichiometric matrix
+            species_solver.pop(ix)
+
+    return species_solver
+
+
+def _get_species_compartment_symbol(species: sbml.Species) -> sp.Symbol:
+    """
+    Generate compartment symbol for the compartment of a specific species.
+    This function will always return the same unique python object for a
+    given species name.
+
+    :param species:
+        sbml species
+    :return:
+        compartment symbol
+    """
+    return sp.Symbol(species.getCompartment(), real=True)
+
+
+def _get_identifier_symbol(var: sbml.SBase) -> sp.Symbol:
+    """
+    Generate identifier symbol for a sbml variable.
+    This function will always return the same unique python object for a
+    given entity.
+
+    :param var:
+        sbml variable
+    :return:
+        identifier symbol
+    """
+    return sp.Symbol(var.getId(), real=True)
+
+
+def _get_species_initial(species: sbml.Species) -> sp.Expr:
+    """
+    Extract the initial concentration from a given species
+
+    :param species:
+        species index
+
+    :return:
+        initial species concentration
+    """
+    amount = species.getInitialAmount()
+
+    # default (allows override from rules)
+    conc = _get_identifier_symbol(species)
+
+    # defined concentration
+    conc_from_conc = sp.sympify(species.getInitialConcentration())
+    # computed concentration
+    conc_from_amount = \
+        sp.sympify(amount) / _get_species_compartment_symbol(species)
+
+    if species.getHasOnlySubstanceUnits():
+        if species.isSetInitialAmount() and not math.isnan(amount):
+            conc = conc_from_amount
+
+        if species.isSetInitialConcentration():
+            conc = conc_from_conc
+
+    else:
+        if species.isSetInitialConcentration():
+            conc = conc_from_conc
+
+        if species.isSetInitialAmount() and not math.isnan(amount):
+            conc = conc_from_amount
+
+    return conc
```

Comment:
```suggestion
    if species.isSetInitialConcentration():
        return sp.sympify(species.getInitialConcentration())
    
    if species.isSetInitialAmount():
        amt = species.getInitialAmount()
        if not math.isnan(amt):
            return sp.sympify(amt) / _get_species_compartment_symbol(species)

    return _get_identifier_symbol(species)
```

Not sure I got the logic, but the code that is evaluated for both conditions of the `if-else` statement looks the same to me? Also, SBML spec says `initialAmount` and `initialConcentration` should not both be specified, so order shouldn't matter right?

Reply:
Yes, was also wondering why this code was necessary, and I think I also tried to simplify but it didn't work, which was probably due to other issues. Just tried locally and eveything passed. Thanks!